homeassistant/components/openerz/sensor.py

```diff
@@ -0,0 +1,138 @@
+"""Support for OpenERZ API for Zurich city waste disposal system."""
+from datetime import datetime, timedelta
+import logging
+
+import requests
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA
+from homeassistant.helpers.entity import Entity
+
+_LOGGER = logging.getLogger(__name__)
+SCAN_INTERVAL = timedelta(hours=12)
+
+# Validation of the user's configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required("zip"): cv.positive_int,
+        vol.Required("waste_type", default="waste"): cv.string,
+        vol.Optional("name"): cv.string,
+    }
+)
+
+
+async def async_setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the sensor platform."""
+    openerz_config = {
+        "zip": config["zip"],
+        "waste_type": config["waste_type"],
+        "name": config.get("name"),
+    }
+    add_entities([OpenERZSensor(openerz_config)])
+
+
+class OpenERZSensor(Entity):
+    """Representation of a Sensor."""
+
+    def __init__(self, openerz_config):
+        """Initialize the sensor."""
+        self._state = None
+        self.openerz_config = openerz_config
+        self.zip = self.openerz_config["zip"]
+        self.waste_type = self.openerz_config["waste_type"]
+        self.friendly_name = self.openerz_config.get("name", self.waste_type)
+        self.start_date = datetime.now()
+        self.end_date = None
+        self.last_api_response = None
+
+        self.update()
+
+    def update_start_date(self):
+        """Set the start day to today."""
+
+        self.start_date = datetime.now()
+
+    def find_end_date(self, day_offset=31):
+        """Find the end date for the request, given an offset expressed in days.
+
+        Args:
+        day_offset (int): difference in days between start and end date of the request
+        """
+
+        self.end_date = self.start_date + timedelta(days=day_offset)
+
+    def make_api_request(self):
+        """Construct a request and send it to the OpenERZ API."""
+
+        headers = {"accept": "application/json"}
+
+        start_date = self.start_date.strftime("%Y-%m-%d")
+        end_date = self.end_date.strftime("%Y-%m-%d")
+
+        payload = {
+            "zip": self.zip,
+            "start": start_date,
+            "end": end_date,
+            "offset": 0,
+            "limit": 0,
+            "lang": "en",
+            "sort": "date",
+        }
+        url = f"http://openerz.metaodi.ch/api/calendar/{self.waste_type}.json"
+
+        try:
+            self.last_api_response = requests.get(url, params=payload, headers=headers)
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on pypi.

https://developers.home-assistant.io/docs/creating_platform_code_review#5-communication-with-devicesservices